README.md
[trybuild]: https://github.com/dtolnay/trybuild
[macrotest]: https://github.com/eupn/macrotest

Reuse(Struct, Enum)
=============

<!-- Include donate button if people find it useful later. -->

[![Build Status](https://travis-ci.org/steadylearner/born.svg?branch=master)](https://travis-ci.org/steadylearner/born)
[![Image Crate](https://img.shields.io/crates/v/born.svg)](https://crates.io/crates/born)
[![Image Doc](https://img.shields.io/badge/rust-documentation-blue.svg)](https://docs.rs/born/
)

It provides functional macros to reuse fields from [Struct](https://doc.rust-lang.org/std/keyword.struct.html) and [Enum](https://doc.rust-lang.org/std/keyword.enum.html) definition.

```toml
[dependencies]
born = "0.0.1"
```

[![born crate example](https://raw.githubusercontent.com/steadylearner/born/master/unite_rust_and_python_with_born.png)](https://raw.githubusercontent.com/steadylearner/born/master/unite_rust_and_python_with_born.png)

<br>

## Why this library?

You can define common fields in Rust struct and enum once and reuse them to remove code duplication. Use it when you want to reuse the same fields for the structs like the example below.

```rust
use born::{
    nested_macro,
    public_struct,
};

public_struct!(
    pub struct UserBase {
        username: String,
        email: String,
        full_name: Option<String>,
    }
);

UserBase!(
    pub struct UserIn {
        pub password: String,
    }
);

// Reuse with the same fields.
UserBase!(
    pub struct UserOut
);

UserBase!(
    pub struct UserInDB {
        pub hashed_password: String,
    }
);
```

Compare it with Python code below from [FAST API](https://fastapi.tiangolo.com/tutorial/extra-models/#reduce-duplication) that inspired this library.

```py
from pydantic import BaseModel, EmailStr

class UserBase(BaseModel):
    username: str
    email: EmailStr
    full_name: str = None


class UserIn(UserBase):
    password: str

# Reuse with the same fields.
class UserOut(UserBase):
    pass


class UserInDB(UserBase):
    hashed_password: str
```

You can see almost same thing is done here to remove code duplication in both parts.

But, different from Python, there is no inheritance of fields with macros from **born**. It is lazily built(born) by your first struct or enum definition.

Everything made from them are completely irrelevant to each other execpt they share the same definition. There is no memory share or something like that.

The macros from this library are **lazy struct and enum builders** to remove code duplication. It is possible with the power of the Rust macro.

## Examples

Here, macros to build public struct and enum are used.

If you want to build private struct and enum, just use macros that start with private and shouldn't use `pub` inside.

### Struct

Say you build a simple demo web server to send private messages.

```rust
use born::{
    nested_macro,
    public_struct,
};

public_struct!(
    // pub is required here before struct
    pub struct MessageBase {
        pub text: String
        // pub text: String // , is not required for the struct definition.
    }
);

MessageBase!(
    #[derive(Debug, Clone, PartialEq)]
    pub struct Message {
        pub read: bool,
        // read: bool, // pub is optional.
    }
);

impl Message {
    fn update_text(&mut self, new_message: String) {
        self.text = new_message
    }
    fn read(&mut self) {
        if self.read == false {
            self.read = true;
        }
    }
}

MessageBase!(
    #[derive(Debug, Clone, PartialEq)]
    pub struct MessageCreateRequest
);

MessageBase!(
    // #[derive(Debug, Clone, PartialEq)]
    pub struct MessageUpdateRequest
);

fn main() {
    let message_create_request = MessageCreateRequest {
        text: "I am Steadylearner and 'born' is the crate name.".into(),
    };

    let mut message = Message {
        text: message_create_request.text,
        read: false,
    };
    println!("{:#?}", &message);

    assert_eq!(message, message.clone());

    let message_update_request = MessageUpdateRequest {
        text: "Reuse fields with macros from 'born'.".into(),
    };

    message.update_text(message_update_request.text);
    println!("{:#?}", &message);

    message.read();
    println!("{:#?}", &message);
}
```

### Enum

[Compare it with the code example from the Rust documenation.](https://doc.rust-lang.org/stable/rust-by-example/custom_types/enum.html)

```rust
use born::{
    nested_macro,
    private_enum,
};

private_enum!(
    enum WebEventBase {
        PageLoad,
        PageUnload, // , here is required if you want to extend it.
    }
);

WebEventBase!(
    // #[derive(Debug, Clone, PartialEq)]
    enum WebEvent {
        KeyPress(char),
        Click { x: i64, y: i64 },
        Paste(String),
    }
);

fn inspect(event: WebEvent) {
    match event {
        WebEvent::PageLoad => println!("page loaded"),
        WebEvent::PageUnload => println!("page unloaded"),
        WebEvent::KeyPress(c) => println!("pressed '{}'.", c),
        WebEvent::Paste(s) => println!("pasted \"{}\".", s),
        WebEvent::Click { x, y } => {
            println!("clicked at x={}, y={}.", x, y);
        },
    }
}

fn main() {
    let pressed = WebEvent::KeyPress('x');
    let pasted  = WebEvent::Paste("my text".to_owned());
    let click   = WebEvent::Click { x: 20, y: 80 };
    let load    = WebEvent::PageLoad;
    let unload  = WebEvent::PageUnload;

    inspect(pressed);
    inspect(pasted);
    inspect(click);
    inspect(load);
    inspect(unload);
}
```

<br>

## Details

- Each struct and enum created from the macros are completely unrelevant to each other except they are built(born) from the same definition.

- When you use `private_struct!` and `private_enum!`, you can't use pub keyword in it and others use them. [It wouldn't be logical if a private struct or private enum can have public fields.](https://doc.rust-lang.org/book/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.html#making-structs-and-enums-public)

- `nested_macro!` is required to use the other macros from this crate. It is used to make a macro that creates other macros.

```rust
macro_rules! nested_macro {
    ($($body:tt)*) => {
        macro_rules! __nested_macro { $($body)+ }
        __nested_macro!($);
    }
}
```

<br>

## Why not attribute macro?

- [You can reuse the fields with attribute macro also.](https://github.com/steadylearner/born-attribute) But, you need some dependencies.

- [If you want more, please read the official documenation about procedural macros.](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros)

<br>

## How to test it

```console
$git clone git@github.com:steadylearner/born.git && cargo test pass
```

1. `$cargo test pass` to run passing tests.
2. `$cargo test fail` to run failing tests. You need to install [trybuild] first.

If you want to see how the macros from this package expand, use `$cargo test macros`. You need to install [rustfmt](https://github.com/rust-lang/rustfmt) and [cargo-expand](https://github.com/dtolnay/cargo-expand) to use it before.

```console
$rustup component add rustfmt && cargo install cargo-expand
```

[macrotest] is based on [trybuild]. They are not that compatible to test with a single command and take long time.

They make cargo to redownload the dependendencies and recompile everytime. For that reason, there are commands to test them separately.

#### License

<sup>
Licensed under either of <a href="LICENSE-APACHE">Apache License, Version
2.0</a> or <a href="LICENSE-MIT">MIT license</a> at your option.
</sup>

<br>

<sub>
Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in this crate by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.
</sub>


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this
file. This change log follows the conventions of
[keepachangelog.com](http://keepachangelog.com/).

## [v0.0.1] - 2020-05-25
### Edited
- [Make born-attribute example to compare with this library.](https://github.com/steadylearner/born-attribute)
- Improve documentations.

## [0.0.0] - 2020-05-22
### Test release of born



tests/pass/mod.rs
// $cargo test pass

pub mod private_enum;
pub mod private_struct;
pub mod public_enum;
pub mod public_struct;

pub mod readme;
pub mod rename;

tests/pass/readme/mod.rs
pub mod r#enum;
pub mod r#struct;
pub mod rename;

tests/pass/readme/rename.rs
use born::{
    nested_macro,
    public_struct,
};

public_struct!(
    pub struct UserBase {
        pub first_name: String,
        pub last_name: String,
        pub email: String,
        pub password: String,
    }
);

UserBase!(
    #[derive(Debug)]
    pub struct User {
        pub id: i64,
    }
);

UserBase!(
    pub struct NewUser
);

// $cargo test -- --nocapture
#[test]
fn pass_readme_rename() {
    let new_user = NewUser {
        first_name: "steady".into(),
        last_name: "learner".into(),
        email: "example@email.com".into(),
        password: "password".into(),
    };

    let user = User {
        id: 0,
        first_name: new_user.first_name,
        last_name: new_user.last_name,
        email: new_user.email,
        password: new_user.password,
    };

    println!("{:#?}", user);
}


src/struct/public.rs
// https://doc.rust-lang.org/rustdoc/documentation-tests.html#documenting-macros

/// Similar to `private_struct!` but public.
///
/// # Example
///
/// ```
/// # #[macro_use] extern crate born;
/// # fn main() {
/// public_struct!(
///     // pub is required before 'struct' when you use public_struct!
///     pub struct MessageBase {
///         pub text: String
///         // text: String // pub is optional in fields.
///     }
/// ); // It is lazy. Nothing is made yet.
///
/// MessageBase!(); // You have to call it to use the struct.
/// let message = MessageBase {
///     text: "First Message".into(),
/// };
/// println!("{}", &message.text);
/// # }
/// ```
#[macro_export]
macro_rules! public_struct {
    (pub struct $commonstruct:ident { $( $commonfieldpub:vis $commonfield:ident: $commonty:ty ),+ $(,)* }) => {
        nested_macro! {
            ($s:tt) => {
                macro_rules! $commonstruct {
                    () => {
                        pub struct $commonstruct {
                            $( $commonfieldpub $commonfield: $commonty, )+
                        }
                    };
                    (#[derive($s($arg:tt)+)]) => {
                        #[derive($s($arg)+)]
                        pub struct $commonstruct {
                            $( $commonfieldpub $commonfield: $commonty, )+
                        }
                    };

                    (pub struct $name:ident { $s( $pub:vis $field:ident: $ty:ty ),+ $s(,)* }) => {
                        pub struct $name {
                            $( $commonfieldpub $commonfield: $commonty, )+
                            $s( $pub $field: $ty ),+
                        }
                    };
                    (#[derive($s($arg:tt)+)] pub struct $name:ident { $s( $pub:vis $field:ident: $ty:ty ),+ $s(,)* }) => {
                        #[derive($s($arg)+)]
                        pub struct $name {
                            $( $commonfieldpub $commonfield: $commonty, )+
                            $s( $pub $field: $ty ),+
                        }
                    };

                    (pub struct $name:ident) => {
                        pub struct $name {
                            $( $commonfieldpub $commonfield: $commonty, )+
                        }
                    };
                    (#[derive($s($arg:tt)+)] pub struct $name:ident) => {
                        #[derive($s($arg)+)]
                        pub struct $name {
                            $( $commonfieldpub $commonfield: $commonty, )+
                        }
                    };
                }
            }
        }
    };
}


src/struct/private.rs
/// Use it to to create, extend and reuse fields from private struct definition.
///
/// # Example
///
/// ```
/// # #[macro_use] extern crate born;
/// # fn main() {
/// private_struct!(
///     struct MessageBase {
///         text: String
///     }
/// ); // It is lazy. Nothing is made yet.
///
/// MessageBase!(); // You have to call it to use the struct.
/// let message = MessageBase {
///     text: "First Message".into(),
/// };
/// println!("{}", &message.text);
/// # }
/// ```
#[macro_export]
macro_rules! private_struct {
    (struct $commonstruct:ident { $( $commonfield:ident: $commonty:ty ),+ $(,)* }) => {
        nested_macro! {
            ($s:tt) => {
                macro_rules! $commonstruct {
                    () => {
                        struct $commonstruct {
                            $( $commonfield: $commonty, )+
                        }
                    };
                    (#[derive($s($arg:tt)+)]) => {
                        #[derive($s($arg)+)]
                        struct $commonstruct {
                            $( $commonfield: $commonty, )+
                        }
                    };

                    (struct $name:ident { $s( $field:ident: $ty:ty ),+ $s(,)* }) => {
                        struct $name {
                            $( $commonfield: $commonty, )+
                            $s( $field: $ty ),+
                        }
                    };
                    (#[derive($s($arg:tt)+)] struct $name:ident { $s( $field:ident: $ty:ty ),+ $s(,)* }) => {
                        #[derive($s($arg)+)]
                        struct $name {
                            $( $commonfield: $commonty, )+
                            $s( $field: $ty ),+
                        }
                    };

                    (struct $name:ident) => {
                        struct $name {
                            $( $commonfield: $commonty, )+
                        }
                    };
                    (#[derive($s($arg:tt)+)] struct $name:ident) => {
                        #[derive($s($arg)+)]
                        struct $name {
                            $( $commonfield: $commonty, )+
                        }
                    };
                }
            }
        }
    };
}


src/lib.rs
// https://doc.rust-lang.org/1.7.0/book/using-rust-without-the-standard-library.html
#![no_std]

#![doc(html_favicon_url = "https://avatars0.githubusercontent.com/u/32325099?s=460&u=cd848fc83d9739939a4ea2d38108c8bcee199109&v=4")]
#![doc(html_logo_url = "https://avatars0.githubusercontent.com/u/32325099?s=460&u=cd848fc83d9739939a4ea2d38108c8bcee199109&v=4")]

#![warn(missing_docs)]
// #![warn(missing_doc_code_examples)] // Use it if necessary.

//! [trybuild]: https://github.com/dtolnay/trybuild
//! [macrotest]: https://github.com/eupn/macrotest
//!
//! Reuse(Struct, Enum)
//! =============
//!
//! It provides functional macros to reuse fields from [Struct](https://doc.rust-lang.org/std/keyword.struct.html) and [Enum](https://doc.rust-lang.org/std/keyword.enum.html) definition.
//!
//! ```toml
//! [dependencies]
//! born = "0.0.0"
//! ```
//!
//! <br>
//!
//! ## Examples
//!
//! ### Struct
//!
//! Say you build a demo web server to send private messages.
//!
//! ```
//! # #[macro_use] extern crate born;
//! use born::{
//!    nested_macro,
//!    public_struct,
//! };
//!
//! // Define your base public struct here.
//! public_struct!(
//!     // pub is required here before struct
//!     pub struct MessageBase {
//!         pub text: String
//!         // pub text: String // , is not required for the struct definition.
//!     }
//! );
//!
//! MessageBase!(
//!     #[derive(Debug, Clone, PartialEq)]
//!     pub struct Message {
//!         pub read: bool,
//!         // read: bool, // pub is optional.
//!     }
//! );
//!
//! impl Message {
//!     fn update_text(&mut self, new_message: String) {
//!         self.text = new_message
//!     }
//!     fn read(&mut self) {
//!         if self.read == false {
//!            self.read = true;
//!        }
//!    }
//! }
//!
//! MessageBase!(
//!    #[derive(Debug, Clone, PartialEq)]
//!    pub struct MessageCreateRequest
//! );
//!
//! MessageBase!(
//!     // #[derive(Debug, Clone, PartialEq)]
//!     pub struct MessageUpdateRequest
//! );
//!
//! fn main() {
//!     let message_create_request = MessageCreateRequest {
//!         text: "I am Steadylearner and 'born' is the crate name.".into(),
//!     };
//!
//!     let mut message = Message {
//!         text: message_create_request.text,
//!         read: false,
//!     };
//!     println!("{:#?}", &message);
//!
//!     assert_eq!(message, message.clone());
//!
//!     let message_update_request = MessageUpdateRequest {
//!         text: "Reuse fields with macros from 'born'.".into(),
//!     };
//!
//!     message.update_text(message_update_request.text);
//!     println!("{:#?}", &message);
//!
//!     message.read();
//!     println!("{:#?}", &message);
//! }
//! ```
//!
//! ### Enum
//!
//! [Compare it with the code example from the Rust documenation for Enum.](https://doc.rust-lang.org/stable/rust-by-example/custom_types/enum.html)
//!
//! ```
//! # #[macro_use] extern crate born;
//! use born::{
//!     nested_macro,
//!     private_enum,
//! };
//!
//! // Define your base private enum here.
//! private_enum!(
//!     enum WebEventBase {
//!         PageLoad,
//!         PageUnload, // , here is required if you want to extend it.
//!     }
//! );
//!
//! WebEventBase!(
//!     // #[derive(Debug, Clone, PartialEq)]
//!     enum WebEvent {
//!         KeyPress(char),
//!         Click { x: i64, y: i64 },
//!         Paste(String),
//!     }
//! );
//!
//! fn inspect(event: WebEvent) {
//!     match event {
//!         WebEvent::PageLoad => println!("page loaded"),
//!         WebEvent::PageUnload => println!("page unloaded"),
//!         WebEvent::KeyPress(c) => println!("pressed '{}'.", c),
//!         WebEvent::Paste(s) => println!("pasted \"{}\".", s),
//!         WebEvent::Click { x, y } => {
//!             println!("clicked at x={}, y={}.", x, y);
//!         },
//!     }
//! }
//!
//! fn main() {
//!     let pressed = WebEvent::KeyPress('x');
//!     let pasted  = WebEvent::Paste("my text".to_owned());
//!     let click   = WebEvent::Click { x: 20, y: 80 };
//!     let load    = WebEvent::PageLoad;
//!     let unload  = WebEvent::PageUnload;
//!
//!     inspect(pressed);
//!     inspect(pasted);
//!     inspect(click);
//!     inspect(load);
//!     inspect(unload);
//! }
//! ```
//!
//! ### Rename
//!
//! You can rename struct and enum and reuse the exact same fields.
//!
//! ```
//! # #[macro_use] extern crate born;
//! use born::{
//!     nested_macro,
//!     public_struct,
//! };
//!
//! public_struct!(
//!     pub struct UserBase {
//!         pub first_name: String,
//!         pub last_name: String,
//!         pub email: String,
//!         pub password: String,
//!     }
//! );
//!
//! UserBase!(
//!     pub struct User {
//!         pub id: i64,
//!     }
//! );
//!
//! UserBase!(
//!     pub struct NewUser
//! );
//!
//! // It is equal to code it manually.
//!
//! // pub struct User {
//! //     pub id: i64,
//! //     pub first_name: String,
//! //     pub last_name: String,
//! //     pub email: String,
//! //     pub password: String,
//! // }
//!
//! // pub struct NewUser {
//! //     pub first_name: String,
//! //     pub last_name: String,
//! //     pub email: String,
//! //     pub password: String,
//! // }
//! ```
//!
//! <br>
//!
//! ## Details
//!
//! - Each struct and enum created from the macros are completely unrelevant to each other except they have the same fields you define.
//!
//! - When you use `private_struct!` and `private_enum!`, you can't use pub keyword in it and others use them. [It wouldn't be logical if a private struct or private enum can have public fields.](https://doc.rust-lang.org/book/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.html#making-structs-and-enums-public)
//!
//! - `nested_macro!` is required to use the other macros from this crate. It is used to make a macro that creates other macros.
//! ```rust
//! macro_rules! nested_macro {
//!     ($($body:tt)*) => {
//!         macro_rules! __nested_macro { $($body)+ }
//!         __nested_macro!($);
//!     }
//! }
//! ```
//!
//! <br>
//!
//! ## Comparison with attribute macros
//!
//! - [You can reuse the fields with attribute macros also.](https://github.com/steadylearner/born-attribute) But, you need some dependencies.
 But, you need some dependencies.
//!
//! - [If you want more, please read the official documenation about procedural macros.](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros)

mod r#enum;
mod r#struct;

/// You can make a macro that creates other macros with it.
///
/// ## Example
///
/// ```
/// macro_rules! public_struct {
///     (pub struct $commonstruct:ident { $( $commonfieldpub:vis $commonfield:ident: $commonty:ty ),+ $(,)* }) => {
///         nested_macro! {
///             ($s:tt) => {
///                 macro_rules! $commonstruct {
///                     () => {
///                         pub struct $commonstruct {
///                             $( $commonfieldpub $commonfield: $commonty, )+
///                         }
///                     };
///                 }
///             }
///         }
///     };
/// }
/// ```
#[macro_export]
macro_rules! nested_macro {
    ($($body:tt)*) => {
        macro_rules! __nested_macro { $($body)+ }
        __nested_macro!($);
    }
}

// It will only work with $cargo test --doc or $cargo test
#[cfg(doctest)]
doc_comment::doctest!("../README.md");
